pelican/utils.py

```diff
@@ -659,79 +659,86 @@ def escape_html(text, quote=True):
-    the same slug have that metadata.
-
-    For each content_list item, sets the 'translations' attribute.
+    `False` or `false` will be used as translations, unless all the items in
+    the same group have that metadata.
+
+    Translations and original items are determined relative to one another
+    amongst items in the same group. Items are in the same group if they
+    have the same value(s) for the metadata attribute(s) specified by
+    'translate_per', which must be a string or a collection of strings.
+    If 'translate_per' is falsy, the identification of translations is skipped
+    and all items are returned as originals.
     """
-    content_list.sort(key=attrgetter('slug'))
-    grouped_by_slugs = groupby(content_list, attrgetter('slug'))
-    index = []
-    translations = []
 
-    def _warn_source_paths(msg, items, *extra):
-        args = [len(items)]
-        args.extend(extra)
-        args.extend((x.source_path for x in items))
-        logger.warning('{}: {}'.format(msg, '\n%s' * len(items)), *args)
+    if not translate_per:
+        return content_list, []
 
-    for slug, items in grouped_by_slugs:
-        items = list(items)
+    if isinstance(translate_per, six.string_types):
+        translate_per = {translate_per}
 
-        # display warnings if slug is empty
-        if not slug:
-            _warn_source_paths('There are %s items with empty slug', items)
-
-        # display warnings if several items have the same lang
-        for lang, lang_items in groupby(items, attrgetter('lang')):
-            lang_items = list(lang_items)
-            if len(lang_items) > 1:
-                _warn_source_paths(
-                    'There are %s items with slug "%s" with lang %s',
-                    lang_items,
-                    slug,
-                    lang)
-
-        # items with `translation` metadata will be used as translations...
-        candidate_items = list(filter(
-            lambda i:
-                i.metadata.get('translation', 'false').lower() == 'false',
-            items))
-        # ...unless all items with that slug are translations
-        if not candidate_items:
-            logger.warning('All items with slug "%s" are translations', slug)
-            candidate_items = items
-
-        # find items with default language
-        original_items = list(filter(
-            attrgetter('in_default_lang'),
-            candidate_items))
-
-        # if there is no article with default language, go back one step
-        if not original_items:
-            original_items = candidate_items
-
-        # display warning if there are several original items
-        if len(original_items) > 1:
-            _warn_source_paths(
-                'There are %s original (not translated) items with slug "%s"',
-                original_items,
-                slug)
+    index = []
 
+    content_list.sort(key=attrgetter(*translate_per))
```

Comment:
Maybe a `try/except` to provide better error messages in case you pass bad attribute names.

Reply:
Messages are good, but you need to raise exceptions (with those messages) instead of logging them. This isn't really a recoverable error, pelican should stop. My intent was to provide more relevant error messages instead of the generic python errors.

```python
    except TypeError:
         raise TypeError('Cannot unpack %s, \'translation_id\' must be falsy, a'
                      'string or a collection of strings' % translation_id)
     except AttributeError:
         raise AttributeError('Cannot use %s as \'translation_id\', there appear to be'
                      'items without these metadata attributes' % translation_id)
```

Reply:
Makes sense, I've changed it to reraise

Reply:
Almost. Unlike logging, this time you're supposed to do the formatting :)

```python
raise TypeError('... %s ...' %s foo)
# or
raise TypeError('... {} ...'.format(foo))
```

Otherwise you're passing a tuple.

Reply:
Doh!